src/test/blockmanager_tests.cpp

```diff
@@ -82,4 +83,116 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_scan_unlink_already_pruned_files, TestChain
     BOOST_CHECK(!AutoFile(OpenBlockFile(new_pos, true)).IsNull());
 }
 
+struct FastPruneTestingSetup : public BasicTestingSetup {
+    FastPruneTestingSetup()
+        : BasicTestingSetup{CBaseChainParams::MAIN, /*extra_args=*/{"-fastprune=1"}} {}
+};
+
+BOOST_FIXTURE_TEST_CASE(blockmanager_flush_block_file, FastPruneTestingSetup)
+{
+    const auto params{CreateChainParams(ArgsManager{}, CBaseChainParams::MAIN)};
+    BlockManager blockman{};
+    CChain chain{};
+
+    // Ensure (by delaying execution of certain steps)
+    // that if a file is modified its timestamp will be updated.
+#ifdef WIN32
+    // FAT filesystem has write time resolution of 2 seconds
+    constexpr std::chrono::milliseconds pause(2000);
+#else
+    // EXT3 and EXT4 have nanosecond resolution
+    constexpr std::chrono::milliseconds pause(1);
+#endif
+
+    // Test blocks with no transactions, not even a coinbase
+    CBlock block1;
+    block1.nVersion = 1;
+    CBlock block2;
+    block2.nVersion = 2;
+    CBlock block3;
+    block3.nVersion = 3;
+
+    // They are 80 bytes header + 1 byte 0x00 for vtx length
+    constexpr int TEST_BLOCK_SIZE{81};
+
+    // Blockstore is empty
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), 0);
+
+    // Write the first block; dbp=nullptr means this block doesn't already have a disk
+    // location, so allocate a free location and write it there.
+    FlatFilePos pos1{blockman.SaveBlockToDisk(block1, /*nHeight=*/1, chain, *params, /*dbp=*/nullptr)};
+
+    // Write second block
+    FlatFilePos pos2 = blockman.SaveBlockToDisk(block2, 2, chain, *params, nullptr);
+
+    // Two blocks in the file
+    BOOST_CHECK_EQUAL(blockman.CalculateCurrentUsage(), (TEST_BLOCK_SIZE + BLOCK_SERIALIZATION_HEADER_SIZE) * 2);
+
+    // First two blocks are written as expected
+    CBlock read_block;
+    node::ReadBlockFromDisk(read_block, pos1, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 1);
+    node::ReadBlockFromDisk(read_block, pos2, params->GetConsensus());
+    BOOST_CHECK_EQUAL(read_block.nVersion, 2);
+
+    // Record last modified time of blk00000.dat
+    const fs::path file_path = node::GetBlockPosFilename(pos2);
+    const fs::file_time_type time1 = fs::last_write_time(file_path);
+
+    std::this_thread::sleep_for(pause);
+
+    // Attempt, but fail, to save block 3 to original position of block 2
+    blockman.SaveBlockToDisk(block3, 3, chain, *params, &pos2);
```

Comment:
I don't understand what this part of the test is testing - calling `SaveBlockToDisk` for a block at a specified but incorrect position is something that should never be possible in an actual, non-corrupted node unless I'm missing something. 
But if it somehow happened anyway, even if the existing block isn't overwritten, I'd imagine that it would still cause some havoc, like changing block file statistics [here](https://github.com/bitcoin/bitcoin/blob/c7f1d95f52883d7087b74f45f5e4ce1100d51149/src/node/blockstorage.cpp#L646-L648) for the added size of a block that is never actually written to disk, which would probably be bad.

Reply:
1. What I'm asserting here is that if a position is passed in, the function called "save block to disk" does not save a block to disk. By using incorrect data I can easily prove that it is not actually written to disk.
2. The call to `AddBlock` you highlight here confuses and concerns me a bit - since we know a `known` block won't be written, why are we updating the metadata at all?

Reply:
> 2\. The call to `AddBlock` you highlight here confuses and concerns me a bit - since we know a `known` block won't be written, why are we updating the metadata at all?

Because the block index LevelDB database (saved in `blocks/index`) is being wiped at the beginning of a reindex and needs to be rebuilt from scratch as we look at each block. Part of this database is meta info about each blockfile ([see bitcoin wiki](https://en.bitcoin.it/wiki/Bitcoin_Core_0.11_(ch_2):_Data_Storage#Block_index_.28leveldb.29)). I think that's the main reason why we have to call `SaveBlockToDisk` for each block during a reindex at all.

Reply:
ok make sense thanks. do you think my first point justifies the method of the test or is that just confusing then and I should remove it?

Reply:
I think that both removing it and adding an explanatory comment would be ok. Maybe the test could also check that the block statistics (`CBlockFileInfo`) is not the same after the call (`nBlocks` should have increased)?

Reply:
added in force-push to 5e579c6435145d84e43fb79224f2cb09b40ea047 and rebased on master